fs/fcb2/include/fcb/fcb.h

```diff
@@ -0,0 +1,214 @@
+    uint16_t si_sector_in_range;    /* Sector number relative to si_range */
+};
+
+/**
+ * Error codes.
+ */
+#define FCB_OK           0
+#define FCB_ERR_ARGS    -1
+#define FCB_ERR_FLASH   -2
+#define FCB_ERR_NOVAR   -3
+#define FCB_ERR_NOSPACE -4
+#define FCB_ERR_NOMEM   -5
+#define FCB_ERR_CRC     -6
+#define FCB_ERR_MAGIC   -7
+#define FCB_ERR_VERSION -8
+
+int fcb_init(struct fcb *fcb);
+
+/*
+ * Initialize fcb for specific flash area
+ *
+ * Function initializes FCB structure with data taken from specified flash
+ * area.
+ * If FCB was not initialized before in this area, area will be erased.
+ *
+ * @param fcb            Fcb to initialize
+ * @param flash_area_id  flash area for this fcb
+ * @param version        version of fcb
+ *
+ */
+int fcb_init_flash_area(struct fcb *fcb, int flash_area_id, uint8_t version);
+
+/**
+ * fcb_log is needed as the number of entries in a log
+ */
+struct fcb_log {
+    struct fcb fl_fcb;
+    uint8_t fl_entries;
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+    /* Internal - tracking storage use */
+    uint32_t fl_watermark_off;
+#endif
+};
+
+/**
+ * fcb_append() appends an entry to circular buffer. When writing the
+ * contents for the entry, fcb_write() with fcb_entry filled by.
+ * fcb_append(). When you're finished, call fcb_append_finish() with
+ * loc as argument.
+ */
+int fcb_append(struct fcb *fcb, uint16_t len, struct fcb_entry *loc);
+int fcb_write(struct fcb_entry *loc, uint16_t off, void *buf, uint16_t len);
+int fcb_append_finish(struct fcb_entry *append_loc);
+
+/**
+ * Walk over all log entries in FCB, or entries in a given flash_area.
+ * cb gets called for every entry. If cb wants to stop the walk, it should
+ * return non-zero value.
+ *
+ * Entry data can be read using fcb_read().
+ */
+typedef int (*fcb_walk_cb)(struct fcb_entry *loc, void *arg);
+int fcb_walk(struct fcb *, int sector, fcb_walk_cb cb, void *cb_arg);
+int fcb_getnext(struct fcb *fcb, struct fcb_entry *loc);
+int fcb_read(struct fcb_entry *loc, uint16_t off, void *buf, uint16_t len);
+
+/**
+ * Erases the data from oldest sector.
+ */
+int fcb_rotate(struct fcb *fcb);
+
+/**
+ * Start using the scratch block.
+ */
+int fcb_append_to_scratch(struct fcb *fcb);
+
+/**
+ * How many sectors are unused.
+ */
+int fcb_free_sector_cnt(struct fcb *fcb);
+
+/**
+ * Whether FCB has any data.
+ */
+int fcb_is_empty(struct fcb *fcb);
+
+/**
+ * Element at offset *entries* from last position (backwards).
+ */
+int
+fcb_offset_last_n(struct fcb *fcb, uint8_t entries,
+        struct fcb_entry *last_n_entry);
+
+/**
+ * Get total size of FCB
+ *
+ * @param fcb     FCB to use
+ *
+ * return FCB's size in bytes
+ */
+int fcb_get_total_size(const struct fcb *fcb);
+
+/**
+ * Erase sector in FCB
+ *
+ * @param fcb     FCB to use
+ * @param sector  sector number to erase 0..f_sector_cnt
+ *
+ * return 0 on success, error code on failure
+ */
+int fcb_sector_erase(const struct fcb *fcb, int sector);
```

Comment:
I'd strongly keeping this function private. I can't see a valid reason for user to erase a sector without going through fcb_rotate().